### PR TITLE
Time-based transaction banning

### PR DESCRIPTION
**Pull request by tomusdrw**
If the pool occupies the queue for more than 5 minutes:
1. Remove it from the pool
2. Ban for next 30 minutes

### CODE REVIEW

File: polkadot/transaction-pool/src/lib.rs [hunk 1]
```diff
@@ -298,37 +323,97 @@ impl<'a, A> txpool::Verifier<UncheckedExtrinsic> for Verifier<'a, A> where
...
+			return false;
+		}
+
+		self.banned_until.write().insert(*tx.hash(), *now + self.ban_time);
+		true
+	}
+
+	/// Removes timed bans.
+	pub fn clear_timeouts(&self, now: &Instant) {
+		let to_remove = {
```
Comment (gavofyork) [inline polkadot/transaction-pool/src/lib.rs]: Why is this in {}s?

File: polkadot/transaction-pool/src/rotator.rs [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+				}
+			}
+		}
+
+		true
+	}
+
+	/// Removes timed bans.
+	pub fn clear_timeouts(&self, now: &Instant) {
+		let to_remove = {
```
Comment (gavofyork) [inline polkadot/transaction-pool/src/rotator.rs]: Why the {}s?

File: polkadot/transaction-pool/src/rotator.rs [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+
+#[cfg(test)]
+mod tests {
+	use super::*;
+	use runtime::{Extrinsic, Call, TimestampCall, UncheckedExtrinsic};
+
+	fn rotator() -> PoolRotator {
+		PoolRotator {
+			ban_time: Duration::from_millis(10),
+			..Default::default()
```
Comment (gavofyork) [inline polkadot/transaction-pool/src/rotator.rs]: Nice :-)

File: polkadot/transaction-pool/src/rotator.rs (hunk 2 above)
Comment (tomusdrw) [reply polkadot/transaction-pool/src/rotator.rs]: We were collecting the keys to remove using only the `.read()` lock. I changed the code though to  acquire the lock only once.

**Pull request merged**